mmedit/models/inpaintors/aot_inpaintor.py

```diff
@@ -0,0 +1,252 @@
+                        metric_name](tensor2img(fake_img, min_max=(-1, 1)),
+                                     tensor2img(gt_img, min_max=(-1, 1)))
+                else:
+                    eval_results[metric_name] = self._eval_metrics[
+                        metric_name]()(data_dict).item()
+            output['eval_results'] = eval_results
+        else:
+            output['fake_res'] = fake_res
+            output['fake_img'] = fake_img
+
+        output['meta'] = None if 'meta' not in kwargs else kwargs['meta'][0]
+
+        if save_image:
+            assert save_image and save_path is not None, (
+                'Save path should been given')
+            assert output['meta'] is not None, (
+                'Meta information should be given to save image.')
+
+            tmp_filename = output['meta']['gt_img_path']
+            filestem = Path(tmp_filename).stem
+            if iteration is not None:
+                filename = f'{filestem}_{iteration}.png'
+            else:
+                filename = f'{filestem}.png'
+            mmcv.mkdir_or_exist(save_path)
+            img_list = [kwargs['gt_img']] if 'gt_img' in kwargs else []
+            img_list.extend(
+                [masked_img,
+                 mask.expand_as(masked_img), fake_res, fake_img])
+            img = torch.cat(img_list, dim=3).cpu()
+            self.save_visualization(img, osp.join(save_path, filename))
+            output['save_img_path'] = osp.abspath(
+                osp.join(save_path, filename))
+
+        return output
+
+    def train_step(self, data_batch, optimizer):
+        """Train step function.
+
+        In this function, the inpaintor will finish the train step following
+        the pipeline:
+        1. get fake res/image
+        2. compute reconstruction losses for generator
+        3. compute adversarial loss for discriminator
+        4. optimize generator
+        5. optimize discriminator
+
+        Args:
+            data_batch (torch.Tensor): Batch of data as input.
+            optimizer (dict[torch.optim.Optimizer]): Dict with optimizers for
+                generator and discriminator (if have).
+        Returns:
+            dict: Dict with loss, information for logger, the number of \
+                samples and results for visualization.
+        """
+        log_vars = {}
+
+        gt_img = data_batch['gt_img']
+        mask = data_batch['mask']
+        masked_img = data_batch['masked_img']
+        masked_img = masked_img.float() + mask
+
+        input_x = torch.cat([masked_img, mask], dim=1)
+        fake_res = self.generator(input_x)
+        fake_img = gt_img * (1. - mask) + fake_res * mask
+
+        # reconstruction losses
+        results, g_losses = self.generator_loss(fake_res, fake_img, data_batch)
+        loss_g, log_vars_g = self.parse_losses(g_losses)
+        log_vars.update(log_vars_g)
+
+        # adversarial loss
+        set_requires_grad(self.disc, True)
```

Comment:
I do not see `set_requires_grad(self.disc, False)`. Is that normal? As I see line 405 [here](self.disc_step_count)

Reply:
I think the reason line 405 sets `set_requires_grad(self.disc, False)` is to reduce the computational overhead when calculating losses for generator. If you look at line 364 `set_requires_grad(self.disc, True)` when the discriminator is trained

Reply:
If we do not set it to False, the discriminator will also have gradient when optimizing the generator. That could lead to incorrect training.

Reply:
I see, I guess I'll follow the `onestageinpaintor` then.

Reply:
Just a sanity check, I have screenshot the official code below. Would be great if you could also compare my `aot_inpaintor` with the official implementation to see if there's any major difference between the two. Thanks!

![image](https://user-images.githubusercontent.com/33197366/150044098-7ed08c61-7491-490a-84cc-950558beac64.png)

I have included the log part below too for your reference:

![image](https://user-images.githubusercontent.com/33197366/150044225-00efb7df-c36a-4c2c-9c96-0de23faffaa9.png)

Reply:
In the official code, adversarial loss is computed using `comp_img`, but the output of the generator is used for computing adversarial loss here?

Reply:
The output of the generator here is `fake_res`, whereas `comp_img` in the official code corresponds to `fake_img` here. 

![Screenshot from 2022-01-20 09-57-45](https://user-images.githubusercontent.com/33197366/150248767-fb0494df-695f-4c2a-a8e6-7851e5943059.png)

`fake_img` is then used to compute the adversarial loss, which is the same input as `comp_img` in the official code:

![Screenshot from 2022-01-20 09-57-56](https://user-images.githubusercontent.com/33197366/150249128-f329311c-c7fc-4502-a84c-9d81a19c8879.png)

Reply:
Oh sorry! I missed that. Then I think there is no problem.